samples/algorithms/sudoku-grover/Program.cs

```diff
@@ -0,0 +1,474 @@
+            var emptySquareEdgesQArray = new QArray<ValueTuple<long, long>>(emptySquareEdges);
+            var startingNumberConstraintsQArray = new QArray<ValueTuple<long, long>>(startingNumberConstraints);
+            Console.WriteLine("Quantum solving puzzle ");
+            ShowGrid(puzzle, size);
+            var task = SolvePuzzle.Run(sim, emptySquares.Count, size, emptySquareEdgesQArray, startingNumberConstraintsQArray);
+            if (!task.Result.Item1)
+                Console.WriteLine("no solution found ");
+            else
+            {
+                var solution = task.Result.Item2.ToArray();
+                for (int i = 0; i < solution.Length; i++)
+                {
+                    puzzle[emptySquares[i].i, emptySquares[i].j] = (int)solution[i] + 1;
+                }
+                Console.WriteLine("solved puzzle ");
+                ShowGrid(puzzle, size);
+            }
+        }
+
+        /// # Summary
+        /// Display the puzzle
+        public static void ShowGrid(int[,] puzzle, int size)
+        {
+            for (int i = 0; i < size; i++)
+            {
+                for (int j = 0; j < size; j++)
+                    Console.Write("----");
+                Console.WriteLine("-");
+                for (int j = 0; j < size; j++) 
+                {
+                    if (puzzle[i, j] == 0)
+                        Console.Write("|   ");
+                    else
+                        Console.Write(String.Format("| {0,1} ", puzzle[i, j]));
+                }
+                Console.WriteLine("|");
+            }
+            for (int j = 0; j < size; j++)
+                    Console.Write("----");
+            Console.WriteLine("-");
+        }
+
+        /// # Summary
+        /// find the puzzle empty square edges, and starting number constraints for those empty squares
+        ///
+        /// # Input
+        /// ## puzzle
+        /// The Sudoku puzzle
+        /// ## size
+        /// The size of the puzzle. For example, 9 for a 9x9 puzzle
+        /// 
+        /// # Output
+        /// ## emptySquareEdges
+        /// The list of empty square edges specifying Vertices in the same row/col/subgrid
+        /// ## startingNumberConstraints
+        /// The set of numbers that this empty square can not have. 
+        /// ## emptySquares
+        /// list of empty squares with their i,j locations
+        public static void FindEdgesAndInitialNumberConstraints(int[,] puzzle, int size, int subSize,
+            out List<ValueTuple<long, long>> emptySquareEdges, out HashSet<ValueTuple<long, long>> startingNumberConstraints,
+            out List<EmptySquare> emptySquares)
+        {
+            // find color edges ... i.e. edges between horizontal, vertical and diagonal empty squares
+            // note that for size=4, we will subtract 1 from all puzzle numbers so that they fit in 2 bits i.e. 1 to 4 becomes 0 to 3
+            int[,] emptyIndexes = new int[size, size];
```

Comment:
```suggestion
            var emptyIndexes = new int[size, size];
```

Reply:
moved this over as a suggestion in the new file.